session.go

```diff
@@ -259,158 +272,231 @@ func (sess *session) Break(c chan<- bool) {
 	sess.breakCh = c
 }
 
-func (sess *session) handleRequests(reqs <-chan *gossh.Request) {
-	for req := range reqs {
-		switch req.Type {
-		case "shell", "exec":
-			if sess.handled {
-				req.Reply(false, nil)
-				continue
-			}
+func (sess *session) handleRequests(ctx Context, reqs <-chan *gossh.Request) {
+	keepAliveEnabled := sess.keepAliveInterval > 0
+	lastReceived := time.Now()
 
-			var payload = struct{ Value string }{}
-			gossh.Unmarshal(req.Payload, &payload)
-			sess.rawCmd = payload.Value
+	var keepAliveCallback func()
 
-			// If there's a session policy callback, we need to confirm before
-			// accepting the session.
-			if sess.sessReqCb != nil && !sess.sessReqCb(sess, req.Type) {
-				sess.rawCmd = ""
-				req.Reply(false, nil)
-				continue
-			}
+	var keepAliveTicker *time.Ticker
+	if keepAliveEnabled {
+		keepAliveTicker = time.NewTicker(sess.keepAliveInterval)
+		defer keepAliveTicker.Stop()
 
-			sess.handled = true
-			req.Reply(true, nil)
+		keepAliveCallback = func() {
+			lastReceived = time.Now()
+			keepAliveTicker.Reset(sess.keepAliveInterval)
+		}
 
-			go func() {
-				sess.handler(sess)
-				sess.Exit(0)
-			}()
-		case "subsystem":
-			if sess.handled {
-				req.Reply(false, nil)
-				continue
-			}
+		ctx.SetValue(ContextKeyKeepAliveCallback, keepAliveCallback)
+	} else {
+		// Configure a stopped ticker to prevent `<-keepAliveTicker.C` from panicking.
+		keepAliveTicker = time.NewTicker(math.MaxInt64)
+		keepAliveTicker.Stop()
+	}
 
-			var payload = struct{ Value string }{}
-			gossh.Unmarshal(req.Payload, &payload)
-			sess.subsystem = payload.Value
+	for {
+		select {
+		case <-keepAliveTicker.C:
+			if lastReceived.Add(time.Duration(sess.keepAliveCountMax) * sess.keepAliveInterval).Before(time.Now()) {
+				log.Println("Keep-alive reply not received. Close down the session.")
 
-			// If there's a session policy callback, we need to confirm before
-			// accepting the session.
-			if sess.sessReqCb != nil && !sess.sessReqCb(sess, req.Type) {
-				sess.rawCmd = ""
-				req.Reply(false, nil)
-				continue
+				err := sess.Close()
+				if err != nil {
+					log.Printf("Closing session failed: %v", err)
+				}
+				return
 			}
 
-			handler := sess.subsystemHandlers[payload.Value]
-			if handler == nil {
-				handler = sess.subsystemHandlers["default"]
+			log.Println("Send keep-alive request to the client")
+			// reply can be either false or true, but it always means that the client is alive
+			_, err := sess.SendRequest(keepAliveRequestType, true, nil)
+			if err != nil {
+				log.Printf("Sending keep-alive request failed: %v", err)
+			} else {
+				log.Println("Client replied to keep-alive request.")
+				keepAliveCallback()
```

Comment:
Perhaps this should call `ctx.KeepAliveCallback()()`?

Reply:
Done